pipelined/bevy_sprite2/src/render/mod.rs

```diff
@@ -146,75 +148,72 @@ impl FromWorld for SpriteShaders {
             let rect = texture_atlas.textures[atlas_sprite.index as usize];
-            extracted_sprites.push(ExtractedSprite {
-                atlas_size: Some(texture_atlas.size),
-                transform: transform.compute_matrix(),
-                rect,
-                handle: texture_atlas.texture.clone_weak(),
-            });
+            commands
+                .get_or_spawn(entity)
+                .insert_bundle((ExtractedSprite {
+                    atlas_size: Some(texture_atlas.size),
+                    transform: transform.compute_matrix(),
+                    rect,
+                    handle: texture_atlas.texture.clone_weak(),
+                    vertex_index: 0,
+                },));
         }
     }
-
-    if let Some(mut extracted_sprites_res) = render_world.get_resource_mut::<ExtractedSprites>() {
-        extracted_sprites_res.sprites.extend(extracted_sprites);
-    }
 }
 
 pub fn extract_sprites(
+    mut commands: Commands,
     images: Res<Assets<Image>>,
-    sprite_query: Query<(&Sprite, &GlobalTransform, &Handle<Image>)>,
-    mut render_world: ResMut<RenderWorld>,
+    sprite_query: Query<(Entity, &Sprite, &GlobalTransform, &Handle<Image>)>,
 ) {
-    let mut extracted_sprites = Vec::new();
-    for (sprite, transform, handle) in sprite_query.iter() {
-        let image = if let Some(image) = images.get(handle) {
-            image
-        } else {
-            continue;
+    let mut sprites = Vec::new();
+    for (entity, sprite, transform, handle) in sprite_query.iter() {
+        if let Some(image) = images.get(handle) {
+            let size = image.texture_descriptor.size;
+
+            sprites.push((
+                entity,
+                (ExtractedSprite {
+                    atlas_size: None,
+                    transform: transform.compute_matrix(),
+                    rect: Rect {
+                        min: Vec2::ZERO,
+                        max: sprite
+                            .custom_size
+                            .unwrap_or_else(|| Vec2::new(size.width as f32, size.height as f32)),
+                    },
+                    handle: handle.clone_weak(),
+                    vertex_index: 0,
+                },),
+            ));
         };
-        let size = image.texture_descriptor.size;
-
-        extracted_sprites.push(ExtractedSprite {
-            atlas_size: None,
-            transform: transform.compute_matrix(),
-            rect: Rect {
-                min: Vec2::ZERO,
-                max: sprite
-                    .custom_size
-                    .unwrap_or_else(|| Vec2::new(size.width as f32, size.height as f32)),
-            },
-            handle: handle.clone_weak(),
-        });
-    }
-
-    if let Some(mut extracted_sprites_res) = render_world.get_resource_mut::<ExtractedSprites>() {
-        extracted_sprites_res.sprites.extend(extracted_sprites);
     }
+    commands.insert_or_spawn_batch(sprites);
```

Comment:
I just want to point out that this can also be done with `commands.get_or_spawn(entity).insert_bundle(bundle)`. I think allocating a temporary `Vec` is unnecessary here, but I don't know the performance implications.

Reply:
If I remember correctly @cart has specifically added the batch spawning to the ECS as an optimisation, because otherwise there is a bunch of redundant work done with spawning individually where the entities get moved between different archetypes. Something like that.

Reply:
Yup using the temporary vec and batch spawning is still waaaay faster (i measured). I do agree that the extra allocation is less than ideal (this same issue applies to the spawn_batch command). I'd like to sort out some way to "auto batch" commands so that we can reuse the space allocated in the command buffer (and potentially make individual `entity(entity).insert(bundle)`-style calls as fast as their batched counterparts. I have thoughts on this ranging from "look ahead when processing the command list for subsequent identical commands" and "commands behave like system-local worlds that get merged into the main world". Both of which will require their own focus / pull requests.

The current solution should be "good enough" for now. If I find time before the 0.6 release I'll try some form of auto-batching, but I can't promise anything :)

Reply:
So `get_or_spawn(entity).insert(component)` is used over and faster `insert_or_spawn_batch` if its no bundle? Or would be `insert_or_spawn_batch` better in any case and we should treat the single components as bundle - `(component,)`? If so there are at least two occurences in which `insert_or_spawn_batch` can be used.

Reply:
No `get_or_spawn(entity).insert(component)` is slower in every case where it is called more than once. The only real benefit it has over batching is "call pattern simplicity". `insert_or_spawn_batch` should be used internally, where possible. I just ported the atlas extraction to use batching, which ill push in a few minutes.